contracts/root/TokenPredicates/BurnableERC20Predicate.sol

```diff
@@ -0,0 +1,115 @@
+        address indexed depositor,
+        address indexed depositReceiver,
+        address indexed rootToken,
+        uint256 amount
+    );
+
+    constructor() public {}
+
+    function initialize(address _owner) external initializer {
+        _setupContractId("BurnableERC20Predicate");
+        _setupRole(DEFAULT_ADMIN_ROLE, _owner);
+        _setupRole(MANAGER_ROLE, _owner);
+    }
+
+    /**
+     * @notice Lock burnable ERC20 tokens for deposit, callable only by manager, when depositing from L1 to L2
+     * @param depositor Address who wants to deposit tokens
+     * @param depositReceiver Address (address) who wants to receive tokens on child chain
+     * @param rootToken Token which gets deposited
+     * @param depositData ABI encoded amount
+     */
+    function lockTokens(
+        address depositor,
+        address depositReceiver,
+        address rootToken,
+        bytes calldata depositData
+    ) external override only(MANAGER_ROLE) {
+        uint256 amount = abi.decode(depositData, (uint256));
+
+        emit LockedBurnableERC20(depositor, depositReceiver, rootToken, amount);
+        IBurnableERC20(rootToken).transferFrom(
+            depositor,
+            address(this),
+            amount
+        );
+    }
+
+    /**
+     * @notice Validates log signature, from and to address
+     * then burns whole withdrawn amount
+     *
+     * @param rootToken L1 token which gets withdrawn
+     * @param log Valid ERC20 burn log from child chain
+     */
+    function exitTokens(
+        address,
+        address rootToken,
+        bytes memory log
+    ) public override only(MANAGER_ROLE) {
+        RLPReader.RLPItem[] memory logRLPList = log.toRlpItem().toList();
+        RLPReader.RLPItem[] memory logTopicRLPList = logRLPList[1].toList(); // topics
+
+        // User who's withdrawing
+        address withdrawer = address(logTopicRLPList[1].toUint());
+
+        // Checking if user wanted to get fund back on L1, after burning it on L2
+        if(bytes32(logTopicRLPList[0].toUint()) == TRANSFER_EVENT_SIG) {
+            
+            require(address(logTopicRLPList[2].toUint()) == address(0), "BurnableERC20Predicate: INVALID_RECEIVER");
+
+            uint256 amount = logRLPList[2].toUint();
+            
+            IBurnableERC20 token = IBurnableERC20(rootToken);
+            token.transfer(withdrawer, amount);
+
+            return;
+
+        }
+
+        // Checking whether user wanted to **actually** burn tokens
+        // on both L2, L1
+        if(bytes32(logTopicRLPList[0].toUint()) == BURN_EVENT_SIG) {
+
+            uint256 amount = logRLPList[2].toUint();
+
+            IBurnableERC20 token = IBurnableERC20(rootToken);
+            // Make sure L1, token burning function can be ( only ) invoked
+            // by this predicate contract
+            //
+            // @note Because this predicate contracts has those tokens locked with self
+            token.burn(address(this), amount);
```

Comment:
Who is expected to call that? there is no incentive for it
I think, as mentioned a better approach is to instead do not use ERC20 transfer event for getting a token transfered to L1 but use a specific ChainExit event that would also allow extra data to go through

Reply:
> Who is expected to call that?

Those who need to call it. I don't see it causing any harm. Extra data for ERC20 not supported.

Reply:
My point was not that it cause harm, but that this event is not necessary.

What matter as discussed on slack is the ability to ensure a token on L2 :
1. can be burnt forever if needed
2. can be burnt to be transfered back to L1
3. that the later can contain extra arbitrary data like but not limited to tokenURI so the L1 contract can be aware of them

The addition of a Burn only bring point 1.

Having a ChainExit event brings all 3
if it is not emitted, then we got 1. (no need to burn it on L1)
if it is emiited we get 1. and 2.

Reply:
> can be burnt forever if needed

It requires burn on L1 too, because in L1 tokens are kept in a single bucket not separately for each user who ever attempted to deposit token.

> can be burnt to be transfered back to L1

This is powering it up : https://github.com/maticnetwork/pos-portal/blob/4bd19325dabfca7d07aa4cf7c46260255a50b66d/contracts/root/TokenPredicates/BurnableERC20Predicate.sol#L83-L94

> that the later can contain extra arbitrary data like but not limited to tokenURI so the L1 contract can be aware of them

As it's ERC20, there's no idea of `tokenURI` in standard, so I've planned adding capability for transferring data from L2 -> L1. But this is possible & probably be done in some future date.

---

**This BurnableERC20 Predicate allows both**

- Burning on L2 & bringing tokens in L1 [ **feature from ERC20Predicate** ]

https://github.com/maticnetwork/pos-portal/blob/4bd19325dabfca7d07aa4cf7c46260255a50b66d/contracts/root/TokenPredicates/BurnableERC20Predicate.sol#L83-L94

- Burn on L2 & do same on L1

https://github.com/maticnetwork/pos-portal/blob/4bd19325dabfca7d07aa4cf7c46260255a50b66d/contracts/root/TokenPredicates/BurnableERC20Predicate.sol#L98-L111

Reply:
> It requires burn on L1 too, because in L1 tokens are kept in a single bucket not separately for each user who ever attempted to deposit token.

No it does not, it can be kept locked forever. The burn is I mentioned unlikely to be triggered in any case as there are no incentive to do it. If we really need that I would use a special operator value to allow the burn to happen on L1 if really required, No need a specific event for that


> As it's ERC20, there's no idea of tokenURI in standard, so I've planned adding capability for transferring data from L2 -> L1. But this is possible & probably be done in some future date.

Of course, but this PR is about ERC1155 , ERC721 and ERC20, and having data being able to be passed through L1 is crucial for many application, including our at Sandbox. would be a shame to builld up this new PR to then have to wait for another to have that in, while the ChainExist event could solve all of that

Reply:
The main issue for me is that the current PR is far more complex that it could.. It add a lot of events while a single event could cater for all what we need with the added advantage that it will support arbitrary data to go through.

Regarding burn, the chainExit could also contain the receiver to the zero address to indicate a burn.

Reply:
> No it does not, it can be kept locked forever. The burn is I mentioned unlikely to be triggered in any case as there are no incentive to do it. If we really need that I would use a special operator value to allow the burn to happen on L1 if really required, No need a specific event for that

It requires burn on L1 too, because in L1 tokens are kept in a single bucket not separately for each user who ever attempted to deposit token - It's not possible to distinguish between who deposited what, when kept all assets in a single bucket.

Reply:
It does not matter if they are instinguishable. If the token are implemented correctly, locked token burn on L2 would not be able to be withdrawn on l1